README.md
# Text2cc – Create quizzes in Common Cartridge format from Markdown-based plain text

Much of the code and documentation are not updated to reflect the new name and Schoology format.
The only question types supported at this time are:

* Multiple Choice
* True / False
* Fill in the blank
* Multiple Answers
* Essay

Some will not work and others I've just not been able to test yet. 

Tested and not Working

* File Upload
* Text Regions
* Question Feedback

 

Text2cc converts
[Markdown](https://daringfireball.net/projects/markdown/)-based plain text
files into quizzes in QTI format (version 1.2), which can be imported by
[Schoology](https://www.schoology.com) and other educational software.
It supports multiple-choice, true/false, multiple-answers, numerical,
short-answer (fill-in-the-blank) and essay questions.  It
includes basic support for LaTeX math within Markdown, and allows a limited
subset of [siunitx](https://ctan.org/pkg/siunitx) notation for units and for
numbers in scientific notation.



## Examples

text2cc allows quick and efficient quiz creation.  Example
**multiple-choice** plain-text quiz question that can be converted to Common Cartridge and
then imported by Schoology:

```
1.  What is 2+3?
a)  6
b)  1
*c) 5
```

A **question** is created by a line that starts with a number followed by a
period and one or more spaces or tabs ("`1. `").  Possible **choices** are
created by lines that start with a letter followed by a closing parenthesis
and one or more spaces or tabs ("`a) `").  Numbers and letters do not have to
be ordered or unique.  The **correct** choice is designated with an asterisk
("`*c) `").  All question and choice text is processed as
[Markdown](https://daringfireball.net/projects/markdown/).

There is also support for a quiz title and description, as well as question
titles, point values, and feedback.  Note that unlike most other text, titles
like quiz and question titles are treated as plain text, not Markdown, due to
the QTI format.  **Also note that Canvas correctly shows question titles
within its quiz editor for instructors, but always replaces them with titles
like "Question 1" in the student view.**  Question point values must be
positive integers or half-integers.

```
Quiz title: Addition
Quiz description: Checking addition.

Title: An addition question
Points: 2
1.  What is 2+3?
... General question feedback.
+   Feedback for correct answer.
-   Feedback for incorrect answer.
a)  6
... Feedback for this particular answer.
b)  1
... Feedback for this particular answer.
*c) 5
... Feedback for this particular answer.
```

**Multiple-answers questions** use `[]` or `[ ]` for incorrect answers and
`[*]` for correct answers.

```
1.  Which of the following are dinosaurs?
[ ] Woolly mammoth
[*] Tyrannosaurus rex
[*] Triceratops
[ ] Smilodon fatalis
```

**Numerical questions** use an equals sign followed by one or
more spaces or tabs followed by the numerical answer.  Acceptable answers can
be designated as a range of the form `[<min>, <max>]` or as a correct answer
with a specified acceptable margin of error `<ans> +- <margin>`.  When the
latter form is used, `<margin>` can be either a number or a percentage.
`<margin>` can be omitted when the answer is an integer and an exact answer is
required.  In this case, scientific notation is not permitted, but the
underscore can be used as a digit separator; for example, `1000` and `1_000`
are both valid, but `1e3` is not.  An exact answer can be required for
floating-point numbers, but this requires an explicit `+- 0`, since a range is
typically more appropriate for floating-point values.  Numerical questions
have the limitation that the absolute value of the smallest acceptable answer
must be greater than or equal to 0.0001 (1e-4).

```
1.  What is the square root of 2?
=   1.4142 +- 0.0001

2.  What is the cube root of 2?
=   [1.2598, 1.2600]

3.  What is 2+3?
=   5
```

**Short-answer (fill-in-the-blank) questions** use an asterisk followed by one
or more spaces or tabs followed by an answer.  Multiple acceptable answers can
be given.  Answers are restricted to a single line each and are treated as
plain text, not Markdown.
```
1.  Who lives at the North Pole?
*   Santa
*   Santa Claus
*   Father Christmas
*   Saint Nicholas
*   Saint Nick
```

**Essay questions** are indicated by a sequence of three or more underscores.
They only support general question feedback.

```
1.  Write an essay.
... General question feedback.
____
```




## Installation

Install **Python 3.6+** if it is not already available on your machine.  See
https://www.python.org/, or use the package manager or app store for your
operating system.  Depending on your use case, you may want to consider a
Python distribution like [Anaconda](https://www.anaconda.com/distribution/)
instead.

Install
[setuptools](https://packaging.python.org/tutorials/installing-packages/)
for Python if it is not already installed.  This can be accomplished by
running
```
python -m pip install setuptools
```
on the command line.  Depending on your system, you may need to use `python3`
instead of `python`.  This will often be the case for Linux and OS X.

Install text2qti by running this on the command line:
```
python -m pip install text2qti
```
Depending on your system, you may need to use `python3` instead of `python`.
This will often be the case for Linux and OS X.


### Upgrading

```
python -m pip install text2qti --upgrade
```
Depending on your system, you may need to use `python3` instead of `python`.
This will often be the case for Linux and OS X.


### Installing the development version

If you want to install the development version to use the latest features,
download `text2qti` from [GitHub](https://github.com/gpoore/text2qti) and
extract the files.  A few different ways to install the development version
are listed below.  Depending on your system, you may need to use `python3`
instead of `python` in the commands below.  This will often be the case for
Linux and OS X.

* You can install using the included `setup.py` by running
  ```
  python setup.py install
  ```
  Depending on your system configuration, especially if you do not have root
  or administrator privileges, you may want to
  [customize the installation location](https://docs.python.org/3.8/install/#alternate-installation-the-user-scheme).
  For example, you can add `--user` to install under `%APPDATA%\Python` (Windows), `~/.local` (UNIX, and Mac OS X non-framework builds), or
  `~/Library/Python/<VERSION>` (Mac framework builds):
  ```
  python setup.py install --user
  ```
* You can install using `pip`.  For example, in the directory with `setup.py`,
  run this:
  ```
  python -m pip install .
  ```




## Usage

text2qti has been designed to create QTI files for use with
[Canvas](https://www.instructure.com/canvas/).  Some features may not be
supported by other educational software.  You should **always preview**
quizzes or assessments after converting them to QTI and importing them.

Write your quiz or assessment in a plain text file.  You can use a basic
editor like Notepad or gedit, or a code editor like
[VS Code](https://code.visualstudio.com/).  You can even use Microsoft Word,
as long as you save your file as plain text (*.txt).

text2qti includes a graphical application and a command-line application.

* To use the graphical application, open a command line and run `text2qti_tk`.

* To use the command-line application, open a command line in the same folder
  or directory as your quiz file.  Under Windows, you can hold the SHIFT
  button down on the keyboard, then right click next to your file, and select
  "Open PowerShell window here" or "Open command window here".  You can also
  launch "Command Prompt" or "PowerShell" through the Start menu, and then
  navigate to your file using `cd`.

  Run the `text2qti` application using a command like this:
  ```
  text2qti quiz.txt
  ```
  Replace "quiz.txt" with the name of your file.  This will create a file like
  `quiz.zip` (with "quiz" replaced by the name of your file) which is the
  converted quiz in QTI format.

Instructions for using the ZIP file with Schoology:

* Go to the resources tab.
* Click the arrow near the top center, then import.
* Select Common Cartridge then choose a new or existing collection and folder name.
* Click Next, then attach the file.
* Select the zip file previously created.
* Click next. Now your quiz should be imported into the folder and collection you chose.
You should **always preview the quiz before use**.  text2CC can detect a
number of potential issues, but not everything.

Typically, you should start your quizzes with a title, like this:
```
Quiz title: Title here
```
Otherwise, all quizzes will have the default title "Quiz", so it will be
difficult to tell them apart.  Another option is to rename quizzes after
importing them.  Note that unlike most other text, the title is treated as
plain text, not Markdown, due to the QTI format.

When you run `text2qti` for the first time, it will attempt to create a
configuration file called `.text2qti.bespon` in your home or user directory.
It will also ask for an institutional LaTeX rendering URL.  This is only
needed if you plan to use LaTeX math; if not, simply press ENTER to continue.
 * If you use Canvas, log into your account and look in the browser address
   bar.  You will typically see an address that starts with something like
   `institution.instructure.com/` or `canvas.institution.edu/`, with
   `institution` replaced by the name of your school or an abbreviation for
   it.  The LateX rendering URL that you want to use will then be something
   like `https://institution.instructure.com/equation_images/` or
   `https://canvas.institution.edu/equation_images/`, with `institution`
   replaced by the appropriate value for your school.  If the URL is like the
   second form, you may need to replace the `.edu` domain with the appropriate
   value for your institution.
 * If you use other educational software that handles LaTeX in a manner
   compatible with Canvas, consult the documentation for your software.  Or
   perhaps create a simple quiz within the software using its built-in tools,
   then export the quiz to QTI and look through the resulting output to find
   the URL.
 * If you are using educational software that does not handle LaTeX in a
   manner compatible with Canvas, try the `--pandoc-mathml` command-line
   option when creating QTI files (note that this requires that
   [Pandoc](https://pandoc.org/) be installed).  If that does not work, please
   open an issue requesting support for that software, and include as much
   information as possible about how that software processes LaTeX.



## Additional features

### Question groups

A question group contains multiple questions, and only a specified number of
these are randomly selected and used each time a quiz is taken.

```
GROUP
pick: 1
points per question: 1

1.  A question.
*a) true
b)  false

2.  Another question.
*a) true
b)  false

END_GROUP
```

The number of questions from the group that are used is specified with
"`pick:`".  If this is omitted, it defaults to `1`.  The points assigned per
question is specified with "`points per question:`".  If this is omitted, it
defaults to `1`.  All questions within a group must be worth the same number
of points.


### Additional quiz options

There are additional quiz options that can be set immediately after the quiz
title and quiz description.  These all take values `true` or `false`.  For
example, `Shuffle answers: true` could be on the line right after the quiz
description.
* `Shuffle answers` — Shuffle answer order for questions.
* `Show correct answers` — Show correct answers after submission.
* `One question at a time` — Only show one question at a time.
* `Can't go back` — Don't allow going back to the previous question when in
  `One question at a time` mode.




## Details for writing quiz text

text2qti processes almost all text as
[Markdown](https://daringfireball.net/projects/markdown/), using
[Python-Markdown](https://python-markdown.github.io/).  (The only exceptions
are the quiz title, question titles, and text region titles, which are
processed as plain text due to the QTI format, plus the acceptable answers
for short-answer questions.)  For example, `*emphasized*` produces *emphasized*
text, which typically appears as italics.  Text can be styled using Markdown
notation, or with HTML.  Remember to preview quizzes after conversion to QTI,
especially when using any significant amount of HTML.

Python-Markdown provides several
[extensions to basic Markdown](https://python-markdown.github.io/extensions/).
Currently, the following extensions are enabled:
* `smarty`:  Automatic curly quotation marks and dashes.
* `sane_lists`:  List behavior is closer to what might be expected.
* `def_list`:  Definition lists of this form:
  ```
  term
  :   definition
* `fenced_code`:  Fenced code blocks (` ``` ` or `~~~`).
* `footnotes`:  Footnotes using this form:
  ```
  Normal text [^1].

  [^1]: Footnote text.
* `tables`:  Tables of this form:
  ```
  Header | Header
  ------ | ------
  Cell   | Cell
  Cell   | Cell
  ```
* `md_in_html`:  Text inside HTML tags is treated as Markdown.  This requires
  setting the attribute `markdown="1"` in the opening tag for block-level
  elements.  See the
  [documentation](https://python-markdown.github.io/extensions/md_in_html/)
  for more details about proper usage and potential issues.

While indented Markdown code blocks are supported, fenced code blocks should
be preferred.  Indented code can interfere with the preprocessor that strips
HTML comments and handles LaTeX math and siunitx notation.


### Titles

Quiz, question, and text region titles are limited to a single paragraph.  If
this paragraph is wrapped over multiple lines, all lines after the first must
be indented by at least two spaces or one tab, and share the same indentation.
All tabs are expanded to 4 spaces before indentation is compared, following
the typical Markdown approach.

All titles are treated as plain text, not Markdown, due to the QTI format.

Titles are always optional, but when they are used for a given element, they
are always required to be first, before any other attributes.


### Descriptions, questions, choices, feedback, and text regions

Descriptions, questions, choices, feedback, and text regions may span multiple
paragraphs and include arbitrary Markdown content like code blocks or
quotations.  Everything must be indented to at least the same level as the
start of the first paragraph on the initial line.  All tabs are expanded to 4
spaces before indentation is compared, following the typical Markdown
approach.  For example,
```
1.  A question paragraph that is long enough to wrap onto a second line.
    The second line must be indented to match up with the start of the
    paragraph text on the first line.

    Another paragraph.
```
Note that the acceptable answers for short-answer questions are treated as
plain text and limited to a single line, and numerical answers are also
processed specially and limited to a single line.


### Images

Images are included with the standard Markdown syntax:
```
![alt_text](image_file)
```
It will typically be easiest to put your image files in the same folder or
directory as the quiz file, so you can use something like `![alt](image.jpg)`.
However, file paths are supported, including `~` user expansion under all
operating systems.  All image paths not starting with `http://` or `https://`
are assumed to refer to local image files (files on your machine), and will
result in errors if these files are not found.


### LaTeX

By default, text2qti supports LaTeX using a Canvas LaTeX rendering URL.  This
can be set during installation, or by editing the configuration file
`.text2qti.bespon` in your home or user directory.  It is possible to convert
LaTeX to MathML instead with the `--pandoc-mathml` command-line option.  This
requires that [Pandoc](https://pandoc.org/) be installed for converting LaTeX
to MathML.  For example, to create a quiz you might run a command like this:
```
text2qti --pandoc-mathml quiz.txt
```
When `--pandoc-mathml` is used, a cache file `_text2qti_cache.zip` will be
created in the quiz file directory.  This is used to store Pandoc MathML
output to increase performance for long quizzes with lots of math.

text2qti supports inline LaTeX math within dollar signs `$`.  There must be a
non-space character immediately after the opening `$` and immediately before
the closing `$`.  For example, `$F = ma$`.  LaTeX math is limited to what is
supported by Canvas or whatever other educational software you are using.  It
is usually a good idea to preview imported quizzes before assigning them,
because text2qti cannot detect LaTeX incompatibilities or limitations.  There
is currently not support for block LaTeX math; only inline math is supported.

When using Canvas with LaTeX math, be aware that in some cases Canvas's
vertical alignment of math leaves much to be desired.  Sometimes this can be
improved by including `\vphantom{fg}` or `\strut` at the beginning of an
equation.  An alternative is simply to use LaTeX for all question or choice
text (via `\text`, etc.).

text2tqi supports a limited subset of LaTeX
[siunitx](https://ctan.org/pkg/siunitx) notation.  You can use notation like
`\num{1.23e5}` to enter numbers in scientific notation.  This would result in
`1.23×10⁵`.  You can use notation like `\si{m/s}` or `\si{N.m}` to enter

CHANGELOG.md
# Change Log


## v0.6.10 (2021-01)

* Forked and renamed Project.
* Removed use of assessment_meta, left file in repo for reference. 
* Changed imsmanifest for Schoology.
* Removed advanced features not supported or tested on Schoology.  
* Converted question types to fit Schoology format.  
* Added version file to support Windows file version info.
* Added Menu Bar and other GUI Improvements.
    * File Menu with Open, Run, and exit
    * Help Menu with help link and about dialog box.
    * Key Bindings and Keyboard Navigation.


## v0.5.0 (2020-09-28)

* Added `text2qti_tk` executable, which provides a basic graphical user
  interface (GUI) via `tkinter`.  Added build scripts in `make_gui_exe/` for
  creating a standalone GUI executable under Windows with PyInstaller (#27).
* In executable code blocks, `.python` now invokes `python3` on systems where
  `python` is equivalent to `python2` as well as on systems that lack a
  `python` executable.  The README now suggests using `.python3` and
  `.python2` to be more explicit (#22).
* Added a new keyword argument `executable` for executable code blocks, which
  allows a custom executable (including path) to be specified for running
  code.  Added support for periods in code block language attributes, so that
  things like `.python3.8` are now possible.
* Fixed bug caused by swapped identifiers in QTI XML (#18, #19).  Now quiz
  descriptions work with Canvas and question titles appear in Canvas in the
  quiz editor (but not in the student view).   The following options now work
  with Canvas:  `Shuffle answers`, `Show correct answers`,
  `One question at a time`, and `Can't go back`.
* Installation (`setup.py`) now requires `markdown` >= 3.2 to ensure
  compatibility and avoid Markdown parsing bugs fixed in 3.2 (#31).
* Fixed bug that produced incorrect QTI output paths when using quiz files
  outside the current working directory (#28, #29).
* README now covers more options for installing the development version (#20).


## v0.4.0 (2020-07-17)

* Improved preprocessing for siunitx notation, LaTeX math, and HTML comments.
  Fixed catastrophic backtracking in LaTeX math regex (#11).  Added support
  for newlines in HTML comments.  The preprocessor now skips backslash
  escapes, inline code containing newlines, and fenced code blocks (as long as
  they do not start on the same line as a list marker).  The preprocessor now
  handles the escape `\$` itself, since Python-Markdown ignores it (#14).
* Python-Markdown's Markdown-in-HTML extension is now enabled (#13).
* Added quiz options `Shuffle answers`, `Show correct answers`,
  `One question at a time`, and `Can't go back` (#10).  These options are
  apparently ignored by Canvas, but may work with some other systems.
* Revised README to clarify that some features are apparently not supported
  by Canvas (#16).



## v0.3.0 (2020-05-26)

* Added support for multiple-answers questions.
* Added support for short-answer (fill-in-the-blank) questions.
* Added support for file-upload questions.
* Added support for setting question titles and point values (#9).
* Added support for text regions outside questions.
* Added `--pandoc-mathml` command-line option.  This converts LaTeX to MathML
  via Pandoc, rather than using a Canvas LaTeX rendering URL (#4).
* Added support for comments at the top level of quiz files (outside Markdown
  content like questions, choices, or feedback).  HTML comments within
  Markdown are now stripped and no longer appear in the final QTI file (#2).
* For numerical questions, exact answers with allowed margin are now treated
  as exact answers, rather than being converted into ranges of values.  This
  primarily affects how feedback for incorrect answers is worded (#7).
* Essay questions now support general feedback.
* Fixed a bug that prevented incorrect question feedback from working.
* Relaxed indentation requirements for quiz titles spanning multiple lines.
  Indentation for wrapped lines must now be at least 2 spaces or 1 tab, rather
  than being equivalent to that of the first character in the title.
* Fixed a bug that allowed trailing whitespace to cause incorrect indentation
  calculations, resulting in indentation errors in valid quizzes.



## v0.2.0 (2020-04-23)

* Added support for images, using standard Markdown syntax.
* Added support for multi-paragraph descriptions, questions, choices, and
  feedback.
* Added support for essay questions.
* Added support for numerical questions.
* Added support for question groups.  A subset of the questions in a group is
  chosen at random when a quiz is taken.
* Added support for executable code blocks.  These can be used to generate
  questions automatically.  This feature requires the command-line flag
  `--run-code-blocks` or setting `run_code_blocks = true` in the config
  file.
* Quiz titles are now processed as plain text, rather than as Markdown,
  because QTI does not support HTML titles.
* Fixed a bug that prevented source file names from appearing in error
  messages.  Fixed misspelling of "imsmanifest.xml" in QTI output.



## v0.1.0 (2020-04-01)

* Initial release.


text2cc/gui/tk.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021, Dana Lehman
# Copyright (c) 2020, Geoffrey M. Poore
# All rights reserved.
#
# Licensed under the BSD 3-Clause License:
# http://opensource.org/licenses/BSD-3-Clause
#


import os
import pathlib
import shutil
import time
import tkinter as tk
import tkinter.filedialog
import webbrowser
from ..config import Config
from ..err import Text2qtiError
from ..qti import QTI
from ..quiz import Quiz
from .. import version
#from tkinter.ttk import * #possibly not needed



def main():
    config = Config()
    config.load()
    file_name = ''

    window = tk.Tk()
    window.title('text2CC')
    # Bring window to front and put in focus
    window.iconify()
    window.update()
    window.deiconify()

    # Window grid setup
    current_row = 0
    column_count = 4


    header_label = tk.Label(
        window,
        text='text2cc – Create quizzes in Common Cartridge format from Markdown-based plain text',
        font=(None, 16),
    )
    header_label.grid(
        row=current_row, column=0, columnspan=column_count, padx=(30, 30),
        sticky='nsew',
    )
    #current_row += 1
    # header_link_label = tk.Label(
        # window,
        # text='github.com/gpoore/text2qti',
        # font=(None, 14), fg='blue', cursor='hand2',
    # )
    # header_link_label.bind('<Button-1>', lambda x: webbrowser.open_new('https://github.com/gpoore/text2qti'))
    # header_link_label.grid(
        # row=current_row, column=0, columnspan=column_count, padx=(30, 30),
        # sticky='nsew',
    # )
    current_row += 1
    version_label = tk.Label(
        window,
        text=f'Version:   {version.__version__}',
    )
    version_label.grid(
        row=current_row, column=0, columnspan=column_count, padx=(30, 30), pady=(0, 30),
        sticky='nsew',
    )
    current_row += 1


    file_browser_label = tk.Label(
        window,
        text='Quiz file:\n(plain text file)',
        justify='right',
    )
    file_browser_label.grid(
        row=current_row, column=0, padx=(30, 5), pady=(5, 25),
        sticky='nse',
    )
    last_dir = None
    def browse_files():
        nonlocal file_name
        nonlocal last_dir
        if last_dir is None:
            initialdir = pathlib.Path('~').expanduser()
        else:
            initialdir = last_dir
        file_name = tkinter.filedialog.askopenfilename(
            initialdir=initialdir,
            title='Select a quiz file',
            filetypes=[('Quiz files', '*.md;*.txt')],
        )
        if file_name:
            if last_dir is None:
                last_dir = pathlib.Path(file_name).parent
            file_browser_button.config(text=f'"{file_name}"', fg='green')
        else:
            file_browser_button.config(text=f'<none selected>', fg='red')

    file_browser_button = tk.Button(
        window,
        text='<none selected>',
        fg='red',
        command=browse_files,
    )
    file_browser_button.grid(
        row=current_row, column=1, columnspan=column_count-1, padx=(0, 30), pady=(5, 25),
        sticky='nsew',
    )
    file_browser_button.bind('<Return>', lambda e: browse_files())
    current_row += 1

 #Above here set good
    # advanced_options_label = tk.Label(
        # window,
        # text='Advanced options – LaTeX math & executable code',
        # justify='right',
    # )
    # advanced_options_label.grid(
        # row=current_row, column=1, columnspan=2, padx=(0, 0), pady=(5, 5),
        # sticky='nsw',
    # )
    # current_row += 1


    # latex_url_label = tk.Label(
        # window,
        # text='LaTeX math rendering URL:\n(for Canvas and similar systems)',
        # justify='right',
    # )
    # latex_url_label.grid(
        # row=current_row, column=0, padx=(30, 5), pady=(5, 5),
        # sticky='nse',
    # )
    # latex_url_entry = tk.Entry(window, width=100)
    # latex_url_entry.grid(
        # row=current_row, column=1, columnspan=column_count-1, padx=(0, 30), pady=(5, 5),
        # sticky='nsew',
    # )
    # if 'latex_render_url' in config:
        # latex_url_entry.insert(1, f"{config['latex_render_url']}")
    # current_row += 1


    # pandoc_exists = bool(shutil.which('pandoc'))
    # pandoc_mathml_label = tk.Label(
        # window,
        # text='Convert LaTeX math to MathML:\n(requires Pandoc; ignores rendering URL)',
        # justify='right',
    # )
    # if not pandoc_exists:
        # pandoc_mathml_label['fg'] = 'gray'
    # pandoc_mathml_label.grid(
        # row=current_row, column=0, padx=(30, 5), pady=(5, 5),
        # sticky='nse',
    # )
    # pandoc_mathml_bool = tk.BooleanVar()
    # def pandoc_mathml_command():
        # if pandoc_mathml_bool.get():
            # latex_url_label['fg'] = 'gray'
            # latex_url_entry['fg'] = 'gray'
        # else:
            # latex_url_label['fg'] = 'black'
            # latex_url_entry['fg'] = 'black'
    # if pandoc_exists:
        # pandoc_mathml_button = tk.Checkbutton(
            # window,
            # variable=pandoc_mathml_bool,
            # command=pandoc_mathml_command,
        # )
        # pandoc_mathml_bool.set(config['pandoc_mathml'])
    # else:
        # pandoc_mathml_button = tk.Checkbutton(
            # window,
            # state=tk.DISABLED,
        # )
    # pandoc_mathml_button.grid(
        # row=current_row, column=1, sticky='w',
    # )
    # current_row += 1


    def run():
        run_message_text.delete(1.0, tk.END)
        run_message_text['fg'] = 'gray'
        run_message_text.insert(tk.INSERT, 'Starting...')
        run_message_text.update()
        error_message = None
        if not file_name:
            error_message = 'Must select a quiz file'
            run_message_text.delete(1.0, tk.END)
            run_message_text.insert(tk.INSERT, error_message)
            run_message_text['fg'] = 'red'
            return
        #if latex_url_entry.get():
            #config['latex_render_url'] = latex_url_entry.get()
        
        #config['pandoc_mathml'] = pandoc_mathml_bool.get()

        file_path = pathlib.Path(file_name)
        try:
            text = file_path.read_text(encoding='utf-8-sig')  # Handle BOM for Windows
        except FileNotFoundError:
            error_message = f'File "{file_path}" does not exist.'
        except PermissionError as e:
            error_message = f'File "{file_path}" cannot be read due to permission error. Technical details:\n\n{e}'
        except UnicodeDecodeError as e:
            error_message = f'File "{file_path}" is not encoded in valid UTF-8. Technical details:\n\n{e}'
        except Exception as e:
            error_message = f'An error occurred in reading the quiz file. Technical details:\n\n{e}'
        if error_message:
            run_message_text.delete(1.0, tk.END)
            run_message_text.insert(tk.INSERT, error_message)
            run_message_text['fg'] = 'red'
            return
        cwd = pathlib.Path.cwd()
        os.chdir(file_path.parent)
        try:
            quiz = Quiz(text, config=config, source_name=file_path.as_posix())
            qti = QTI(quiz)
            qti.save(f'{file_path.stem}.zip')
        except Text2qtiError as e:
            error_message = f'Quiz creation failed:\n\n{e}'
        except Exception as e:
            error_message = f'Quiz creation failed unexpectedly. Technical details:\n\n{e}'
        finally:
            os.chdir(cwd)
        if error_message:
            run_message_text.delete(1.0, tk.END)
            run_message_text.insert(tk.INSERT, error_message)
            run_message_text['fg'] = 'red'
        else:
            run_message_text.delete(1.0, tk.END)
            run_message_text.insert(tk.INSERT, f'Created quiz "{file_path.parent.as_posix()}/{file_path.stem}.zip"')
            run_message_text['fg'] = 'green'
    run_button = tk.Button(
        window,
        text='RUN',
        font=(None, 14),
        command=run,
    )
    run_button.grid(
        row=current_row, column=1, columnspan=2, padx=(0, 0), pady=(30, 30),
        sticky='nsew',
    )
    run_button.bind('<Return>', lambda e: run())
    current_row += 1


    run_message_label = tk.Label(
        window,
        text='\nRun Summary:\n',
        relief='ridge',
        width=120,
    )
    run_message_label.grid(
        row=current_row, column=0, columnspan=column_count, padx=(30, 30), pady=(0, 0),
        sticky='nsew',
    )
    current_row += 1


    run_message_frame = tk.Frame(
        window,
        width=120, height=40,
        borderwidth=1, relief='sunken', bg='white',
    )
    run_message_frame.grid(
        row=current_row, column=0, columnspan=column_count, padx=(30, 30), pady=(0, 30),
        sticky='nsew',
    )
    run_message_scrollbar = tk.Scrollbar(run_message_frame)
    run_message_scrollbar.pack(
        side='right', fill='y',
    )
    run_message_text = tk.Text(
        run_message_frame,
        width=10, height=10, borderwidth=0, highlightthickness=0,
        wrap='word',
        yscrollcommand=run_message_scrollbar.set,
    )
    run_message_text.insert(tk.INSERT, 'Waiting...')
    run_message_text['fg'] = 'gray'
    run_message_scrollbar.config(command=run_message_text.yview)
    run_message_text.pack(
        side='left', fill='both', expand=True,
        padx=(5, 5), pady=(5, 5),
    )
    
    def About_page():
            
         abt = tk.Toplevel(window)
         def close_about():
            abt.destroy()

         abt.grab_set()
         
         abt.title("About")
          
         lbl=tk.Label(
             abt,
             text ='Text2CC \n \n Copyright © 2021, Dana Lehman \n Copyright © 2020, Geoffrey M. Poore \n  ',
             #anchor='e',
             justify='center',
             
             )
         lbl.grid(
            #  column=0,
            #  columnspan=4,
            #  row=1,
            #  rowspan=4,
              pady=3,
              padx=30,
            #  sticky="NW"
             )

         vlbl=tk.Label(
             abt,
             text =f'Version:   {version.__version__} \n',
             #bg = 'red',
             #anchor="CENTER",
             #width=max((len(text)))
            )
         vlbl.grid(
            #  column=0,
            #  columnspan=4,
            #  row=2,
            #  rowspan=4,
              pady=3,
            #  padx=30,
            #  sticky="NW"
             )  

         liclbl = tk.Label(
                abt,
                text='License: BSD 3-Clause', 
                #font=(None, 14), 
                fg='blue', cursor='hand2',
                )
         liclbl.bind('<Button-1>', lambda x: webbrowser.open_new('https://opensource.org/licenses/BSD-3-Clause'))
         liclbl.grid(
                row=current_row, column=0, columnspan=column_count, padx=(30, 30),
                #sticky='nsew',
            )  

         OK_BTN=tk.Button(
             abt,
             text="OK",
             command=close_about,
             anchor= 'center',
            )
         OK_BTN.bind('<Return>', lambda e: close_about())
         OK_BTN.grid(
            #  column=2,
            #  columnspan=1,
            #  row=5,
            #  padx=30,
              pady=30,
            #  sticky="se"
             )
         OK_BTN.focus_set()
        
         # Gets the requested values of the height and widht.
         windowWidth = abt.winfo_reqwidth()
         windowHeight = abt.winfo_reqheight()

         # Gets both half the screen width/height and window width/height
         positionRight = int(abt.winfo_screenwidth()/2 - windowWidth/2)
         positionDown = int(abt.winfo_screenheight()/2 - windowHeight/2)
 
         # Positions the window in the center of the page.
         abt.geometry("+{}+{}".format(positionRight, positionDown))
         #abt.geometry("225x100+{}+{}".format(positionRight, positionDown))
         abt.resizable(height= None, width= None)
         abt.focus()
         
         #abt.bind('<Return>', close_about)
    
    def open_help():
        webbrowser.open_new('https://github.com/dlehman83/text2cc/blob/master/README.md')

    menubar = tk.Menu(window)
    filemenu = tk.Menu(menubar, tearoff=False)
    filemenu.add_command(label="Open",command=browse_files,underline=0)
    filemenu.add_command(label="Run",command=run,underline=0)
    filemenu.add_command(label="Exit",command=window.quit,underline=1)
    menubar.add_cascade(label="File", menu=filemenu,underline=0)
    helpmenu = tk.Menu(menubar, tearoff=False)
    helpmenu.add_command(label="Help",underline=0,command=open_help)
    helpmenu.add_command(label="About",command=About_page,underline=0)
    menubar.add_cascade(label="Help", menu=helpmenu,underline=0)

    # Gets the requested values of the height and widht.
    windowWidth = window.winfo_reqwidth()
    windowHeight = window.winfo_reqheight()

    # Gets both half the screen width/height and window width/height
    positionRight = int(window.winfo_screenwidth()/2 - windowWidth/2)
    positionDown = int(window.winfo_screenheight()/2 - windowHeight/2)
 
    # Positions the window in the center of the page.
    window.geometry("+{}+{}".format(positionRight, positionDown))
    window.config(menu=menubar)
    window.bind('<Return>',run)
    window.mainloop()
